src/tracer/MOM_neutral_diffusion.F90

```diff
@@ -266,44 +279,107 @@ subroutine neutral_diffusion_calc_coeffs(G, GV, h, T, S, EOS, CS)
+  else
+    if (.not. ALLOCATED(T_i))     ALLOCATE(T_i(SZI_(G),SZJ_(G),SZK_(G),2))
+    if (.not. ALLOCATED(S_i))     ALLOCATE(S_i(SZI_(G),SZJ_(G),SZK_(G),2))
+    if (.not. ALLOCATED(dRdT_i))  ALLOCATE(dRdT_i(SZI_(G),SZJ_(G),SZK_(G),2))
+    if (.not. ALLOCATED(dRdS_i))  ALLOCATE(dRdS_i(SZI_(G),SZJ_(G),SZK_(G),2))
+    if (.not. ALLOCATED(ppoly_coefficients))  ALLOCATE(ppoly_coefficients(SZK_(G),3))
+
+    T_i(:,:,:,:) = 0.
+    S_i(:,:,:,:) = 0.
+    dRdT_i(:,:,:,:) = 0.
+    dRdS_i(:,:,:,:) = 0.
+    ppoly_coefficients(:,:) = 0.
+  endif
 
   do j = G%jsc-1, G%jec+1
     ! Interpolate state to interface
     do i = G%isc-1, G%iec+1
-      call interface_scalar(G%ke, h(i,j,:), T(i,j,:), Tint(i,j,:), 2)
-      call interface_scalar(G%ke, h(i,j,:), S(i,j,:), Sint(i,j,:), 2)
+      if (CS%continuous_reconstruction) then
+        call interface_scalar(G%ke, h(i,j,:), T(i,j,:), Tint(i,j,:), 2)
+        call interface_scalar(G%ke, h(i,j,:), S(i,j,:), Sint(i,j,:), 2)
+      else
+        call PPM_reconstruction(G%ke, h(i,j,:), T(i,j,:), T_i(i,j,:,:), ppoly_coefficients)
+        if (CS%boundary_extrapolation) then
+          call PPM_boundary_extrapolation(G%ke, h(i,j,:), T(i,j,:), T_i(i,j,:,:), ppoly_coefficients)
+        endif
+        call PPM_reconstruction(G%ke, h(i,j,:), S(i,j,:), S_i(i,j,:,:), ppoly_coefficients)
+        if (CS%boundary_extrapolation) then
+          call PPM_boundary_extrapolation(G%ke, h(i,j,:), S(i,j,:), S_i(i,j,:,:), ppoly_coefficients)
+        endif
+      endif
     enddo
 
     ! Calculate interface properties
     Pint(:,j,1) = 0. ! Assume P=0 (Pa) at surface - needs correcting for atmospheric and ice loading - AJA
-    do k = 1, G%ke+1
-      call calculate_density_derivs(Tint(:,j,k), Sint(:,j,k), Pint(:,j,k), &
-                                    dRdT(:,j,k), dRdS(:,j,k), G%isc-1, G%iec-G%isc+3, EOS)
-      if (k<=G%ke) Pint(:,j,k+1) = Pint(:,j,k) + h(:,j,k) * GV%H_to_Pa ! Pressure at next interface, k+1 (Pa)
-    enddo
+    ! Continuous reconstruction
+    if (CS%continuous_reconstruction) then
+      do k = 1, G%ke
+          call calculate_density_derivs(Tint(:,j,k), Sint(:,j,k), Pint(:,j,k), &
+                                        dRdT(:,j,k), dRdS(:,j,k), G%isc-1, G%iec-G%isc+3, EOS)
+        if (k<=G%ke) then
+          Pint(:,j,k+1) = Pint(:,j,k) + h(:,j,k) * GV%H_to_Pa ! Pressure at next interface, k+1 (Pa)
+        endif
+      enddo
+    else ! Discontinuous reconstruction
+      do k = 1, G%ke
+        call calculate_density_derivs(T_i(:,j,k,1), S_i(:,j,k,1), Pint(:,j,k), &
+                                      dRdT_i(:,j,k,1), dRdS_i(:,j,k,1), G%isc-1, G%iec-G%isc+3, EOS)
+        if (k<=G%ke) then
+          Pint(:,j,k+1) = Pint(:,j,k) + h(:,j,k) * GV%H_to_Pa ! Pressure at next interface, k+1 (Pa)
+          call calculate_density_derivs(T_i(:,j,k,2), S_i(:,j,k,2), Pint(:,j,k+1), &
+                                        dRdT_i(:,j,k,1), dRdS_i(:,j,k,2), G%isc-1, G%iec-G%isc+3, EOS)
```

Comment:
PS> For the record, the use of indices 1,2 for left,right in the reconstruction code is not my doing and I don't like it because you can't read the code. One day we'll clean up the reconstruction interfaces...